### ISSUE TITLE
rocker/rstudio cannot open DESCRIPTION

### ISSUE THREAD

**Issue by wush978** [rocker-org/rocker#84]
Dear All,

First of all, thanks for your work of bringing R to docker.

The rocker/studio container cannot edit the `DESCRIPTION` file. The error message is: `File is binary rather than text so cannot be opened by the source editor.`

Here is how I reproduce this error:
1. `docker run -p 28787:8787 rocker/rstudio`
2. browse `localhost:28787`
3. Create a new project of R package
4. Try to open the file `DESCRIPTION`

Thanks,
Wush

---

According to the wiki, here is some info for you:

OS: ubuntu 14.04 x86_64 
Linux Kernel: 3.13.0-37-generic
Docker version: 1.3.0, build c78088f
The image id: 171039443de1

Docker logs:

```
/usr/lib/python2.7/dist-packages/supervisor/options.py:296: UserWarning: Supervisord is running as root and it is searching for its configuration file in default locations (including its current working directory); you probably want to specify a "-c" argument specifying an absolute path to a configuration file for improved security.
  'Supervisord is running as root and it is searching '
2014-11-14 04:15:38,561 CRIT Supervisor running as root (no user in config file)
2014-11-14 04:15:38,561 WARN Included extra file "/etc/supervisor/conf.d/supervisord.conf" during parsing
2014-11-14 04:15:38,575 INFO RPC interface 'supervisor' initialized
2014-11-14 04:15:38,575 CRIT Server 'unix_http_server' running without any HTTP authentication checking
2014-11-14 04:15:38,576 INFO supervisord started with pid 1
2014-11-14 04:15:39,578 INFO spawned: 'rserver' with pid 12
2014-11-14 04:15:39,578 INFO spawned: 'userconf' with pid 13
2014-11-14 04:15:39,584 INFO exited: rserver (exit status 0; not expected)
2014-11-14 04:15:40,721 INFO spawned: 'rserver' with pid 58
2014-11-14 04:15:40,721 INFO success: userconf entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2014-11-14 04:15:40,722 INFO exited: userconf (exit status 0; expected)
2014-11-14 04:15:40,728 INFO exited: rserver (exit status 0; not expected)
2014-11-14 04:15:41,218 CRIT reaped unknown pid 59)
2014-11-14 04:15:43,221 INFO spawned: 'rserver' with pid 78
2014-11-14 04:15:43,233 INFO exited: rserver (exit status 0; not expected)
2014-11-14 04:15:43,712 CRIT reaped unknown pid 79)
2014-11-14 04:15:46,717 INFO spawned: 'rserver' with pid 98
2014-11-14 04:15:46,725 INFO exited: rserver (exit status 0; not expected)
2014-11-14 04:15:47,192 INFO gave up: rserver entered FATAL state, too many start retries too quickly
2014-11-14 04:15:47,192 CRIT reaped unknown pid 99)
```

**Comment by snim2** [comment]
I think this bug may have reappeared. @sje30 has a repository with a Dockerfile here:

https://github.com/sje30/eglen2015/blob/master/Dockerfile

and instructions on running that Dockerfile here:

https://github.com/sje30/eglen2015#docker

@sje30 can run this container fine on a 1.4.X  version of Docker, but using version 1.5.0 on Ubuntu 15.04 I get this error from RStudio:

    21 Apr 2015 15:29:03 [rsession-rstudio] ERROR system error 28 (No space left on device) [path=/home/rstudio]; OCCURRED AT: core::Error core::system::file_monitor::<unnamed>::addWatch(const core::FileInfo&, const core::FilePath&, bool, int, core::system::file_monitor::<unnamed>::Watches*) /home/ubuntu/rstudio/src/cpp/core/system/file_monitor/LinuxFileMonitor.cpp:225; LOGGED FROM: core::Error session::modules::files::FilesListingMonitor::start(const core::FilePath&, core::json::Array*) /home/ubuntu/rstudio/src/cpp/session/modules/SessionFilesListingMonitor.cpp:66

(my HDD has >700GB free BTW) and this error from the log of the Docker container:

    2015-04-21 15:27:53,464 CRIT Supervisor running as root (no user in config file)
    2015-04-21 15:27:53,465 INFO supervisord started with pid 1
    2015-04-21 15:27:54,467 INFO spawned: 'rserver' with pid 8
    2015-04-21 15:27:54,468 INFO spawned: 'userconf' with pid 9
    2015-04-21 15:27:54,589 INFO success: userconf entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
    2015-04-21 15:27:54,589 INFO exited: rserver (exit status 0; not expected)
    2015-04-21 15:27:55,713 INFO spawned: 'rserver' with pid 35
    2015-04-21 15:27:55,719 INFO exited: rserver (exit status 0; not expected)
    2015-04-21 15:27:56,107 CRIT reaped unknown pid 36)
    2015-04-21 15:27:57,789 INFO spawned: 'rserver' with pid 61
    2015-04-21 15:27:57,793 INFO exited: rserver (exit status 0; not expected)
    2015-04-21 15:27:58,176 INFO exited: userconf (exit status 0; expected)
    2015-04-21 15:27:58,179 CRIT reaped unknown pid 63)
    2015-04-21 15:28:01,183 INFO spawned: 'rserver' with pid 84
    2015-04-21 15:28:01,189 INFO exited: rserver (exit status 0; not expected)
    2015-04-21 15:28:01,581 INFO gave up: rserver entered FATAL state, too many start retries too quickly
    2015-04-21 15:28:01,581 CRIT reaped unknown pid 85)

I added `sudo apt-get install file` etc. to the Dockerfile that @sje30 wrote and oddly that didn't fix the bug there. However, that Dockerfile inherits from `rocker/hadleyverse` and I got confused by how all the `rocker-org` repos relate to each other, so I'm not sure whether the PR mentioned on this issue *should* have fixed the bug or not.  

There are a few questions on SO about similar problems. Answers suggest looking at the config files for `supervisord` and some answers suggest passing the `-n` flag which (apparently) runs `supervisord` in the foreground. I notice that there are several `supervisord` config files in the `rocker-org` repos, but I haven't tried patching them.

**Comment by cboettig** [comment]
@snim2 

- The error messages you see in the logs from supervisord don't have anything to do with this.  Sorry for not suppressing these messages, but they are just known bugs in how supervisord handles exit statuses in the background, see https://github.com/Supervisor/supervisor/issues, and not causing any trouble.

- You seem to be having trouble running an image that we don't maintain, and which doesn't appear in any of our images.  I suggest you file a bug report in the repo of the Dockerfile that is creating the error.

- You've filed this under a closed bug that has nothing to do with your error (yes, I realize that isn't always obvious with these cryptic error messages, but just letting you know).  This thread and the issue about installing 'file' has to do with RStudio-server being able to open files that had no standard extension on them, like "DESCRIPTION" as opposed to "script.R".  

- @sje30 's container works exactly as advertised for me, so the problem isn't on his side either.

The error message you are getting,

```
21 Apr 2015 15:29:03 [rsession-rstudio] ERROR system error 28 (No space left on device) 
```

is coming from Docker, though I'm not sure the context.  Docker does weird things with disk space, particularly on older kernels, so you might want to verify that you do indeed have the free disk space needed. e.g. can you pull down the latest hadleyverse image and run everything fine in that?

**Comment by sje30** [comment]
Thanks @snim2  and @cboettig for trying to track this down.  I think the error Sarah found also manifests itself with the following

    docker run -d -p 8787:8787 rocker/rstudio

Apologies if this is not the right place to report it, but as Sarah kindly agreed to help me debug, I suggested she send some data.

**Comment by cboettig** [comment]
@sje30 @snim2 My suspicion is that this error message is correct and that her machine does not have the free space, so it doesn't matter what the container is.  

For instance, try just pulling a few images:

```
docker pull rocker/hadleyverse
docker pull ruby
```
 
and see if that works.

**Comment by snim2** [comment]
Sigh. Sorry if this is the wrong place to report the problem, it just looked like the same bug. Again, definitely not a lack of space:

    $ df -h
    Filesystem      Size  Used Avail Use% Mounted on
    /dev/sda6       900G   71G  784G   9% /

I tried `sudo docker run -d -p 8787:8787 rocker/hadleyverse` and that has the same problem. 

@cboettig are you using docker `1.5.0`?

**Comment by sje30** [comment]
@snim2 Can you check inodes `df -i`?

**Comment by snim2** [comment]
Filesystem        Inodes  IUsed     IFree IUse% Mounted on
    /dev/sda6       59908096 656248  59251848    2% /

**Comment by wch** [comment]
One more thought - perhaps try using Docker 1.6.0? That's what I have on my Ubuntu 14.04 system. It's installed not via the regular Ubuntu repositories, but from the repository run by Docker. (The Docker package is named `lxc-docker`, as opposed to the package from Ubuntu, which is named `docker.io`.)

**Comment by cboettig** [comment]
@snim2 I'm on docker 1.6.  Yeah, weird on the space.  did you try running those `docker pull` commands?

**Comment by snim2** [comment]
Did I do a `docker pull`? I've never used the `rocker-org` before yesterday, so I just did a `docker run` and let docker download your images from the Hub.

Trying docker `1.6` sounded sensible, so I did a `sudo apt-get purge` on docker, added the relevant PPA and installed `lxc-docker`. `sudo docker version` then returned:

    Client version: 1.6.0
    Client API version: 1.18
    Go version (client): go1.4.2
    Git commit (client): 4749651
    OS/Arch (client): linux/amd64
    Server version: 1.6.0
    Server API version: 1.18
    Go version (server): go1.4.2
    Git commit (server): 4749651
    OS/Arch (server): linux/amd64

and I got the same problem with `rocker/hadleyverse` and `rocker/rstudio`. 

I then purged `lxc-docker` and installed version `1.4.1` from the same PPA, so `sudo docker version` said this:

    Client version: 1.4.1
    Client API version: 1.16
    Go version (client): go1.3.3
    Git commit (client): 5bc2ff8
    OS/Arch (client): linux/amd64
    Server version: 1.4.1
    Server API version: 1.16
    Go version (server): go1.3.3
    Git commit (server): 5bc2ff8

and the problem still persisted. 

Then I thought maybe docker is just completely b0rked on Ubuntu 15.04 and the problem isn't with rocker at all, so I tried `sudo docker run -ti ubuntu:latest bash` and that worked just fine on Docker `1.4.1` and `1.5`. 

At this stage I'm all out of ideas :(

**Comment by nsymms** [comment]
Not sure if I should jump in here, but I get the "no space left" message from time to time when I'm running the ropensci container. At first I tried all sorts of things but eventually gave up. It didn't cause me any problems other than the "file" window in RStudio wouldn't auto-update to any changes in the file system. I could create, edit, run, and otherwise do everything else in RStudio without problems.

The issue has to do with the iNotify part of Linux filesystems and how it works with Docker. The "no space left" error is just a generic message and as @snim2 mentions, probably doesn't mean you're out of storage. Like I said, I never figured out what the problem is exactly; I got tired of looking. It's intermittent for me anyway.

Sorry to further pollute an already closed issue, but I just wanted to hopefully clear some of the fog.

**Comment by cboettig** [comment]
@snim2  There can be a lot going on with filesystem handling in docker that has nothing to do with the free space shown by `df`.  Please Google your error or just browse the issues list in docker and you can find plenty of examples, e.g. https://github.com/docker/docker/issues/10613

I suggested `docker pull ruby` in order to help isolate the issue.  Yes, I know that `run` pulls images you don't have, but it makes it harder to know at what stage we are getting the error.  I also suggested pulling an image you probably don't have to see if docker had space to pull a new image, and to see if this error could be reproduced with a non-rocker container, since it sounds to me like you have an issue with your docker installation that is not particular to the rocker container.  If you had ever pulled the ubuntu container before, then it would run just fine in your example without running into a space issue, so it doesn't test that.

**Comment by snim2** [comment]
@cboettig apologies, I didn't report back on everything I did very precisely. I /did/ also pull `rstudio` after doing a plain `run` and `hadleyverse`, I did pull `ubuntu:latest` and I had not pulled that repo before. I have just this minute pulled the `ruby` container (for the first time), and the ruby interpreter works fine, I can create new files inside the Ubuntu container, etc. So, no problems with pulling or with the other images, AFAICT.

I looked through the Docker issues list, there are about 10 pages of issues relating to "out of space", but I couldn't find anything obvious that would fix this issue, neither did I find anything on Google or Stack Overflow when I looked last night. 

Since you cannot reproduce this bug I'm happy to test anything else you suggest if you think it is useful to rocker, but since I'm not an R user I'm not sure how much help I can realistically be. If @nsymms thinks this bug doesn't affect the usability of the containers, then perhaps you would rather write this one off as a wont-fix. 

Sorry if you feel I've wasted your time.
